File: C++/Localization.cpp
#include "Localization.h"
#include <fstream>
#include <stack>
using namespace SimpleLion;

Localization::Localization(std::string path) {
	_basepath = new boost::filesystem::path(path.c_str());
	// If directory does not exist, create one. This will certainly help our user.
	if(!boost::filesystem::exists(*_basepath)) {
		if(!boost::filesystem::create_directories(*_basepath)) {
			throw SimpleLion::FilesystemException("Could not create directory!", _basepath->string());
		}	
	}
	_locale = "";
	_file = NULL;
	_fileSize = 0;
	// "Open" the base directory.
	boost::filesystem::directory_iterator begin(*_basepath);
	boost::filesystem::directory_iterator end;
	// Directory should not be empty, because we are loading stuff.
	if(begin == end) {
		throw SimpleLion::FilesystemException("No files present!", _basepath->string());
	}
	// How many SLLF files are there in the directory?
	_localesSize = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_localesSize++;
		}
	}
	// Store locale ids (the stem() returns that) into the array.
	if(_localesSize == 0) {
		throw SimpleLion::FilesystemException("No SLLF files present!", _basepath->string());
	}
	_locales = new std::string[_localesSize];
	int i = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_locales[i] = (*iter).path().stem().string();
			i++;
		}
	}
}

Localization::Localization(std::string path, std::string locale) {
	_basepath = new boost::filesystem::path(path.c_str());
	_file = NULL;
	_fileSize = 0;
	// If directory does not exist, create one. This will certainly help our user.
	if(!boost::filesystem::exists(*_basepath)) {
		if(!boost::filesystem::create_directories(*_basepath)) {
			throw SimpleLion::FilesystemException("Could not create directory!", _basepath->string());
		}	
	}
	
	// "Open" the base directory.
	boost::filesystem::directory_iterator begin(*_basepath);
	boost::filesystem::directory_iterator end;
	// Directory should not be empty, because we are loading stuff.
	if(begin == end) {
		throw SimpleLion::FilesystemException("No files present!", _basepath->string());
	}
	// How many SLLF files are there in the directory?
	_localesSize = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_localesSize++;
		}
	}
	if(_localesSize == 0) {
		throw SimpleLion::FilesystemException("No SLLF files present!", _basepath->string());
	}
	// Store locale ids (the stem() returns that) into the array.
	_locales = new std::string[_localesSize];
	int i = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_locales[i] = (*iter).path().stem().string();
			i++;
		}
	}
	// Check if selected locale file is available
	bool available = false;
	for(int j = 0; j < _localesSize; j++) {
		if(_locales[j] == locale) {
			available = true;
			break;
		}
	}
	if(available) {
		_locale = locale;
	} else {
		throw SimpleLion::FileException("Selected locale is unavailable!");
	}
	
	// Ok, so here we know that the locale exists and also where it is. Time to parse it...
	
	boost::filesystem::path localepath = *_basepath;
	localepath += "/";
	localepath += locale;
	localepath += ".sllf";
	
	// The "for sure" check
	
	if(!boost::filesystem::exists(localepath)) {
		throw SimpleLion::FileException("File does not exist!", localepath.string());
	}
	
	// Parse file to structures
	std::ifstream sllfin(localepath.c_str());
	if(!sllfin) {
		throw SimpleLion::FileException("Could not open file!", localepath.string());
	}
	SimpleLion::LocalizationCategory* cat = NULL;
	std::stack<SimpleLion::LocalizationCategory*> catstack;
	std::string name = "", value = "";
	int lineno = 0;
	while(!sllfin.eof()) {
		char c;
		std::string line;
		lineno++;
		while(c != '\n') {
			c = sllfin.get();
			if(sllfin) {
				if(c != '\n') {
					line += c;
				}
			} else {
				break;
			}
		}
		if(sllfin) {
			if(line[0] == '[') {
				std::string parsed = "";
				bool correct = true;
				for(int i = 1; line[i] != ']'; i++) {
					if(i == line.length()) {
						parsed = "";
						correct = false;
						break;
					} else {
						parsed += line[i];
					}
				}
				if(name != "") {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Not adding previous string because found the beginning of another string declaration!" << std::endl;
				}
				if(parsed == "" || !correct) {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Skipping invalid/empty string declaration!" << std::endl;
				}
				name = parsed;
			} else if(line[0] == '{') {
				std::string parsed = "";
				bool correct = true;
				for(int i = 1; line[i] != '}'; i++) {
					if(i == line.length()) {
						parsed = "";
						correct = false;
						break;
					} else {
						parsed += line[i];
					}
				}
				if(parsed == "" || !correct) {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Skipping invalid/empty category declaration!" << std::endl;
				} else {
					if(parsed == "end") {
						if(cat == NULL) {
							std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Encountered end-of-category not matching to a category! Skipping..." << std::endl;
						} else {
							if(!catstack.empty()) {
								SimpleLion::LocalizationCategory* tmpcat = catstack.top();
								catstack.pop();
								tmpcat->addLocalizationEntry(*cat);
								cat = tmpcat;
								tmpcat = NULL;
							} else {
								int newSize = _fileSize+1;
								SimpleLion::LocalizationEntry** newFile = new SimpleLion::LocalizationEntry*[newSize];
								if(_file != NULL) {
									for(int i = 0; i < _fileSize; i++) {
										newFile[i] = _file[i];
									}
									delete[] _file;
								}
								newFile[newSize-1] = cat;
								_fileSize = newSize;
								_file = newFile;
								cat = NULL;
							}
						}
					} else {
						if(cat == NULL) {
							cat = new SimpleLion::LocalizationCategory(parsed);
						} else {
							catstack.push(cat);
							cat = new SimpleLion::LocalizationCategory(parsed);
						}
					}
				}
			} else {
				value = line;
				if(name == "") {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno <<  " WARNING: Skipping value of invalid/empty string declaration!" << std::endl;
				} else {
					if(value == "") {
						std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Skipping empty value of string!" << std::endl;
					} else {
						SimpleLion::LocalizationString* ls = new SimpleLion::LocalizationString(name, value);
						if(cat == NULL) {
							int newSize = _fileSize+1;
							SimpleLion::LocalizationEntry** newFile = new SimpleLion::LocalizationEntry*[newSize];
							if(_file != NULL) {
								for(int i = 0; i < _fileSize; i++) {
									newFile[i] = _file[i];
								}
								delete[] _file;
							}
							newFile[newSize-1] = ls;
							_fileSize = newSize;
							_file = newFile;
						} else {
							cat->addLocalizationEntry(*ls);
						}
					}
					name = "";
					value = "";
				}
			}
		}
	}
	sllfin.close();
	// We' ve got the file parsed and loaded (or so...). Or the exception has been thrown outside.`
}

Localization::Localization(Localization& cpy) {
	_basepath = new boost::filesystem::path(*(cpy._basepath));
	_locale = cpy._locale;
	_fileSize = cpy._fileSize;
	_localesSize = cpy._localesSize;
	if(cpy._file == NULL) {
		_file = cpy._file;
	} else {
		_file = new SimpleLion::LocalizationEntry*[cpy._fileSize];
		for(int i = 0; i < _fileSize; i++) {
			_file[i] = cpy._file[i];
		}
	}
	if(cpy._locales == NULL) {
		_locales = cpy._locales;
	} else {
		_locales = new std::string[_localesSize];
		for(int i = 0; i < _localesSize; i++) {
			_locales[i] = cpy._locales[i];
		}
	}
}

Localization::~Localization() {
	if(_basepath != NULL) {
		delete _basepath;
	}
	_basepath = NULL;
	_locale = "";
	if(_file != NULL) {
		for(int i = 0; i < _fileSize; i++) {
			delete _file[i];
		}
		delete[] _file;
	}
	_fileSize = 0;
	_file = NULL;
	if(_locales != NULL) {
		delete[] _locales;
	}
	_localesSize = 0;
	_locales = NULL;
}

void Localization::setLocale(std::string locale) {
	// Clear structures if necessary
	if(_file != NULL) {
		for(int i = 0; i < _fileSize; i++) {
			delete _file[i];
		}
		delete[] _file;
		_file = NULL;
		_fileSize = 0;
	}
	// "Open" the base directory.
	boost::filesystem::directory_iterator begin(*_basepath);
	boost::filesystem::directory_iterator end;
	// Directory should not be empty, because we are loading stuff.
	if(begin == end) {
		throw SimpleLion::FilesystemException("No files present!", _basepath->string());
	}
	// How many SLLF files are there in the directory?
	_localesSize = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_localesSize++;
		}
	}
	if(_localesSize == 0) {
		throw SimpleLion::FilesystemException("No SLLF files present!", _basepath->string());
	}
	// Store locale ids (the stem() returns that) into the array.
	_locales = new std::string[_localesSize];
	int i = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_locales[i] = (*iter).path().stem().string();
			i++;
		}
	}
	// Check if selected locale file is available
	bool available = false;
	for(int j = 0; j < _localesSize; j++) {
		if(_locales[j] == locale) {
			available = true;
			break;
		}
	}
	if(available) {
		_locale = locale;
	} else {
		throw SimpleLion::FileException("Selected locale is unavailable!");
	}
	
	// Ok, so here we know that the locale exists and also where it is. Time to parse it...
	
	boost::filesystem::path localepath = *_basepath;
	localepath += "/";
	localepath += locale;
	localepath += ".sllf";
	
	// The "for sure" check
	
	if(!boost::filesystem::exists(localepath)) {
		throw SimpleLion::FileException("File does not exist!", localepath.string());
	}
	
	// Parse file to structures
	std::ifstream sllfin(localepath.c_str());
	if(!sllfin) {
		throw SimpleLion::FileException("Could not open file!", localepath.string());
	}
	SimpleLion::LocalizationCategory* cat = NULL;
	std::stack<SimpleLion::LocalizationCategory*> catstack;
	std::string name = "", value = "";
	int lineno = 0;
	while(!sllfin.eof()) {
		char c = ' ';
		std::string line;
		lineno++;
		while(c != '\n') {
			c = sllfin.get();
			if(sllfin) {
				if(c != '\n') {
					line += c;
				}
			} else {
				break;
			}
		}
		if(sllfin) {
			if(line[0] == '[') {
				std::string parsed = "";
				bool correct = true;
				for(int i = 1; line[i] != ']'; i++) {
					if(i == line.length()) {
						parsed = "";
						correct = false;
						break;
					} else {
						parsed += line[i];
					}
				}
				if(name != "") {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Not adding previous string because found the beginning of another string declaration!" << std::endl;
				}
				if(parsed == "" || !correct) {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Skipping invalid/empty string declaration!" << std::endl;
				}
				name = parsed;
			} else if(line[0] == '{') {
				std::string parsed = "";
				bool correct = true;
				for(int i = 1; line[i] != '}'; i++) {
					if(i == line.length()) {
						parsed = "";
						correct = false;
						break;
					} else {
						parsed += line[i];
					}
				}
				if(parsed == "" || !correct) {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Skipping invalid/empty category declaration!" << std::endl;
				} else {
					if(parsed == "end") {
						if(cat == NULL) {
							std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Encountered end-of-category not matching to a category! Skipping..." << std::endl;
						} else {
							if(!catstack.empty()) {
								SimpleLion::LocalizationCategory* tmpcat = catstack.top();
								catstack.pop();
								tmpcat->addLocalizationEntry(*cat);
								cat = tmpcat;
								tmpcat = NULL;
							} else {
								int newSize = _fileSize+1;
								SimpleLion::LocalizationEntry** newFile = new SimpleLion::LocalizationEntry*[newSize];
								if(_file != NULL) {
									for(int i = 0; i < _fileSize; i++) {
										newFile[i] = _file[i];
									}
									delete[] _file;
								}
								newFile[newSize-1] = cat;
								_fileSize = newSize;
								_file = newFile;
								cat = NULL;
							}
						}
					} else {
						if(cat == NULL) {
							cat = new SimpleLion::LocalizationCategory(parsed);
						} else {
							catstack.push(cat);
							cat = new SimpleLion::LocalizationCategory(parsed);
						}
					}
				}
			} else {
				value = line;
				if(name == "") {
					std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno <<  " WARNING: Skipping value of invalid/empty string declaration!" << std::endl;
				} else {
					if(value == "") {
						std::cout << "SLLF Parser: " << localepath.filename() << ":" << lineno << " WARNING: Skipping empty value of string!" << std::endl;
					} else {
						SimpleLion::LocalizationString* ls = new SimpleLion::LocalizationString(name, value);
						if(cat == NULL) {
							int newSize = _fileSize+1;
							SimpleLion::LocalizationEntry** newFile = new SimpleLion::LocalizationEntry*[newSize];
							if(_file != NULL) {
								for(int i = 0; i < _fileSize; i++) {
									newFile[i] = _file[i];
								}
								delete[] _file;
							}
							newFile[newSize-1] = ls;
							_fileSize = newSize;
							_file = newFile;
						} else {
							cat->addLocalizationEntry(*ls);
						}
					}
					name = "";
					value = "";
				}
			}
		}
	}
	sllfin.close();
	// We' ve got the file parsed and loaded (or so...). Or the exception has been thrown outside.
}

std::string Localization::query(std::string queryString) {
	SimpleLion::LocalizationCategory* cat = NULL;
	std::string parsed = "";
	if(_file == NULL) {
		return "simplelion:err:no_locale";
	} else {
		for(int i = 0; i < queryString.length(); i++) {
			if(queryString[i] == '.') {
				if(parsed == "") {
					if(cat != NULL) {
						delete cat;
					}
					
					std::string msg = "simplelion:err:invalid_query: " + queryString;
					return msg;
				} else {
					if(cat != NULL) {
						SimpleLion::LocalizationEntry& tmp = cat->getLocalizationEntryByName(parsed);
						if(tmp.type() == "category") {
							cat = dynamic_cast<SimpleLion::LocalizationCategory*>(&tmp);
						} else if(tmp.type() == "entry") {
							std::string msg = "simplelion:err:translation_missing: " + queryString;
							return msg;
						}
					} else {
						bool found = false;
						for(int j = 0; j < _fileSize; j++) {
							SimpleLion::LocalizationEntry *tmp = _file[j];
						if(tmp->type() == "category" && tmp->name() == parsed) {
							cat = dynamic_cast<SimpleLion::LocalizationCategory*>(tmp);
							found = true;
							break;
						}
						}
						if(!found) {
							std::string msg = "simplelion:err:translation_missing: " + queryString;
							return msg;
						}
					}
				}
				parsed = "";
			} else {
				parsed += queryString[i];
			}
		}
		if(parsed == "") {
			std::string msg = "simplelion:err:invalid_query: " + queryString;
			return msg;
		}
		if(cat != NULL) {
			SimpleLion::LocalizationEntry &tmp = cat->getLocalizationEntryByName(parsed);
			if(tmp.type() == "string") {
				return dynamic_cast<SimpleLion::LocalizationString&>(tmp).value();
			} else {
				std::string msg = "simplelion:err:translation_missing: " + queryString;
				return msg;
			}
		} else {
			SimpleLion::LocalizationEntry *tmp;
			for(int i = 0; i < _fileSize; i++) {
				if(_file[i]->name() == parsed && _file[i]->type() == "string") {
					return dynamic_cast<SimpleLion::LocalizationString*>(_file[i])->value();
				}
			}
			std::string msg = "simplelion:err:translation_missing: " + queryString;
			return msg; 
		}                          
	}
}

std::string* Localization::localeList() {
	if(_locales != NULL) {
		delete[] _locales;
		_locales = NULL;
		_localesSize = 0;
	}
	// "Open" the base directory.
	boost::filesystem::directory_iterator begin(*_basepath);
	boost::filesystem::directory_iterator end;
	// Directory should not be empty, because we are loading stuff.
	if(begin == end) {
		throw SimpleLion::FilesystemException("No files present!", _basepath->string());
	}
	// How many SLLF files are there in the directory?
	_localesSize = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_localesSize++;
		}
	}
	if(_localesSize == 0) {
		throw SimpleLion::FilesystemException("No SLLF files present!", _basepath->string());
	}
	// Store locale ids (the stem() returns that) into the array.
	_locales = new std::string[_localesSize];
	int i = 0;
	for(boost::filesystem::directory_iterator iter(*_basepath); iter != end; iter++) {
		if((*iter).path().extension().string() == ".sllf") {
			_locales[i] = (*iter).path().stem().string();
			i++;
		}
	}
	return _locales;
}

int Localization::localeListSize() {
	return _localesSize;
}

std::string Localization::toSLLF() {
	if(_file == NULL) {
		return "";
	} else {
		std::string ret = "";
		for(int i = 0; i < _fileSize; i++) {
			ret += _file[i]->toSLLF();
		}
		return ret;
	}
}

File: Ruby/lib/localization.rb
require 'localization_string'
require 'localization_category'
require 'exceptions'

##
# ==The SimpleLion module for Ruby
#
# SimpleLion is a multi-language, simple localization framework inspired by Ruby on Rails localization engine. This is a language-specific module for Ruby.
#
# The basic information on the project and wiki can be (for now) found at http://redmine.phitherek.mooo.com and you can also contribute for the project there!

module SimpleLion
	
	##
	# This is the main class of the SimpleLion module for Ruby. It handles parsing, storing and querying data from SLLF files.
	
	class Localization
		
		##
		# This constructor initializes Localization object with data from given +path+ and +locale+
		#
		# +locale+ can be nil, only list of available locales is loaded then
		#
		# It raises FilesystemException when there is a problem with the directory (it can' t open or create it or it has no locale files inside)
		#
		# It raises FileException when the selected locale file does not exist
		#
		# The SLLF Parser inside the constructor outputs warnings if the SLLF file format is wrong in some place, it also outputs the line number where the problem exists
		
		def initialize path, locale=nil
			if File.directory?(path)
				@basepath = path
			else
				begin
					Dir.mkdir(path)
				rescue
					raise SimpleLion::FilesystemException.new(path, "Not a valid directory and cannot be created!")
				end
				@basepath = path
			end
			@locales = []
			basedir = Dir.new(@basepath)
			basedir.each do |fn|
				if File.file?(@basepath + "/" + fn) && fn.split('.')[1] == 'sllf'
					@locales << File.basename(fn, ".sllf")
				end
			end
			if @locales.empty?
				raise SimpleLion::FilesystemException.new(@basepath, "No SLLF files present!")
			end
			if locale != nil
				if !@locales.include?(locale)
					raise SimpleLion::FileException.new(@basepath + "/" + locale + ".sllf", "No such locale!")
				end
				@locale = locale
				@file = []
				filepath = @basepath + "/" + @locale + ".sllf"
				File.open(filepath) do |fh|
					name = ""
					value = ""
					category = nil
					categories = []
					i = 0
					fh.readlines.each do |line|
						line.strip!
						i = i+1
						if line[0] == '['
							if !name.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Not adding previous string because found the beginning of another string declaration!\n"
							end
							parsed = line.partition("[").last.partition("]").first
							sep1 = line.partition("[")[1]
							sep2 = line.partition("[").last.partition("]")[1]
							if sep1.empty? || sep2.empty?
								parsed = ""
							end
							if parsed.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping invalid/empty string declaration!\n"
							end
							name = parsed
						elsif line[0] == '{'
							parsed = line.partition("{").last.partition("}").first
							sep1 = line.partition("{")[1]
							sep2 = line.partition("{").last.partition("}")[1]
							if sep1.empty? || sep2.empty?
								parsed = ""
							end
							if parsed.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping invalid/empty category declaration!\n"
							else
								if parsed == "end"
									if category == nil
										puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Encountered end-of-category not matching to a category! Skipping...\n"
									else
										tmpcat = category
										category = categories.pop
										if category == nil
											@file << tmpcat
										else
											category.addLocalizationEntry(tmpcat)
										end

									end
								else
									if category != nil
										categories << category
									end
									category = SimpleLion::LocalizationCategory.new(parsed)
								end
							end
						else
							value = line
							if name.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping value of invalid/empty string declaration!\n"
							else
								if value.empty?
									puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping empty value of string!\n"
								else
									str = SimpleLion::LocalizationString.new(name, value)
									if category == nil
										@file << str
									else
										category.addLocalizationEntry(str)
									end
								end
							end
							name = ""
							value = ""
						end
					end
				end
			end
		end
		
		##
		# This method loads or changes locale in the Localization object. The SLLF locale file is loaded, parsed and stored inside the object
		#
		# It raises FilesystemException when there are no locales in the base directory of the Localization object
		#
		# It raises FileException when the selected locale file does not exist
		#
		# The SLLF Parser inside the method outputs warnings if the SLLF file format is wrong in some place, it also outputs the line number where the problem exists
		
		def setLocale locale
			@locales = []
			basedir = Dir.new(@basepath)
			basedir.each do |fn|
				if File.file?(@basepath + "/" + fn) && fn.split('.')[1] == 'sllf'
					@locales << File.basename(fn, ".sllf")
				end
			end
			if @locales.empty?
				raise SimpleLion::FilesystemException.new(@basepath, "No SLLF files present!")
			end
			if locale != nil
				if !@locales.include?(locale)
					raise SimpleLion::FileException.new(@basepath + "/" + locale + ".sllf", "No such locale!")
				end
				@locale = locale
				@file = []
				filepath = @basepath + "/" + @locale + ".sllf"
				File.open(filepath) do |fh|
					name = ""
					value = ""
					category = nil
					categories = []
					i = 0
					fh.readlines.each do |line|
						line.strip!
						i = i+1
						if line[0] == '['
							if !name.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Not adding previous string because found the beginning of another string declaration!\n"
							end
							parsed = line.partition("[").last.partition("]").first
							sep1 = line.partition("[")[1]
							sep2 = line.partition("[").last.partition("]")[1]
							if sep1.empty? || sep2.empty?
								parsed = ""
							end
							if parsed.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping invalid/empty string declaration!\n"
							end
							name = parsed
						elsif line[0] == '{'
							parsed = line.partition("{").last.partition("}").first
							sep1 = line.partition("{")[1]
							sep2 = line.partition("{").last.partition("}")[1]
							if sep1.empty? || sep2.empty?
								parsed = ""
							end
							if parsed.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping invalid/empty category declaration!\n"
							else
								if parsed == "end"
									if category == nil
										puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Encountered end-of-category not matching to a category! Skipping...\n"
									else
										tmpcat = category
										category = categories.pop
										if category == nil
											@file << tmpcat
										else
											category.addLocalizationEntry(tmpcat)
										end

									end
								else
									if category != nil
										categories << category
									end
									category = SimpleLion::LocalizationCategory.new(parsed)
								end
							end
						else
							value = line
							if name.empty?
								puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping value of invalid/empty string declaration!\n"
							else
								if value.empty?
									puts "SLLF Parser: " + File.basename(filepath) + ":" + i.to_s + " WARNING: Skipping empty value of string!\n"
								else
									str = SimpleLion::LocalizationString.new(name, value)
									if category == nil
										@file << str
									else
										category.addLocalizationEntry(str)
									end
								end
							end
							name = ""
							value = ""
						end
					end
				end
			end
		end
		
		##
		# This method queries the Localization object for the localized value of the string according to +queryString+
		#
		# The queryString has format: string_name or category_name.category_name.(...).string_name - (...) means n more category names
		#
		# It returns the value of the string found with +queryString+ in current locale or error message:
		# * simplelion:err:invalid_query - if there is a problem with the query
		# * simplelion:err:no_locale - if no locale is loaded
		# * simplelion:err:translation_missing - if no string could be found that matches given +queryString+
		# These error messages are returned as a string value, so they will eventually be displayed in the place of missing/invalid localized strings
		
		def query queryString
			if queryString == nil
				return "simplelion:err:invalid_query: nil"
			end
			if queryString.empty?
				return "simplelion:err:invalid_query: empty"
			end
			splittedQuery = queryString.split('.')
			if @file == nil || @file.empty?
				return "simplelion:err:no_locale"
			end
			cat = nil
			splittedQuery.each do |curr|
				if curr == ""
					return "simplelion:err:invalid_query: " + queryString
				else
					if cat == nil
						found = false
						@file.each do |entry|
							if entry.name.to_s == curr
								if splittedQuery.last == curr
									if entry.is_a?(SimpleLion::LocalizationString)
										return entry.value
									end
								else
									if entry.is_a?(SimpleLion::LocalizationCategory)
										cat = entry
										found = true
									end
								end
							end
						end
						if !found
							return "simplelion:err:translation_missing: " + queryString
						end
					else
						entry = cat.getLocalizationEntryByName curr
						if entry == nil
							return "simplelion:err:translation_missing: " + queryString
						end
						if splittedQuery.last == curr
							if entry.is_a?(SimpleLion::LocalizationString)
								return entry.value
							else
								return "simplelion:err:translation_missing: " + queryString
							end
						else
							if entry.is_a?(SimpleLion::LocalizationCategory)
								cat = entry
							else
								return "simplelion:err:translation_missing: " + queryString
							end
						end
					end
				end
			end
		end
		
		##
		# This method returns an array of available locale codes
		#
		# Every run of this method refreshes the available locale list and so its output is always up-to-date
		#
		# It can also return empty list, but it should not happen because the constructor takes care of locale presence
		
		def localeList
			@locales = []
			basedir = Dir.new(@basepath)
			basedir.each do |fn|
				if File.file?(@basepath + "/" + fn) && fn.split('.')[1] == 'sllf'
					@locales << File.basename(fn, ".sllf")
				end
			end
			@locales
		end
		
		##
		# This method outputs the Localization in the SLLF format - it is identical that parsed file unless the Localization object has been modified in some way in the script
		#
		# It returns nil if no locale file has been loaded
		
		def toSLLF
			if @file == nil || @file.empty?
				nil
			else
				out = ""
				@file.each do |entry|
					out += entry.toSLLF
				end
				out
			end
		end
	end
end
